core-plugins/src/main/java/co/cask/hydrator/plugin/batch/source/XMLReaderBatchSource.java

```diff
@@ -0,0 +1,305 @@
+    if (StringUtils.isNotEmpty(config.tableName)) {
+      pipelineConfigurer.createDataset(config.tableName, KeyValueTable.class.getName());
+    }
+  }
+
+  @Override
+  public void prepareRun(BatchSourceContext context) throws Exception {
+    Job job = JobUtils.createInstance();
+    Configuration conf = job.getConfiguration();
+    if (StringUtils.isNotEmpty(config.pattern)) {
+      conf.set(XMLInputFormat.XML_INPUTFORMAT_PATTERN, config.pattern);
+    }
+    XMLInputFormat.addInputPaths(job, config.path);
+    conf.set(XMLInputFormat.XML_INPUTFORMAT_REPROCESSING_REQUIRED, this.config.reprocessingReq);
+    //set file tracking information in a temporary file, to be available to read outside plugin.
+    setFileTrackingInfo(context, conf);
+    if (StringUtils.isNotEmpty(this.config.actionAfterProcess)) {
+      conf.set(XMLInputFormat.XML_INPUTFORMAT_FILE_ACTION, this.config.actionAfterProcess);
+    }
+    if (StringUtils.isNotEmpty(this.config.actionAfterProcess)) {
+      conf.set(XMLInputFormat.XML_INPUTFORMAT_TARGET_FOLDER, this.config.targetFolder);
+    }
+    conf.set(XMLInputFormat.XML_INPUTFORMAT_NODE_PATH, this.config.nodePath);
+    XMLInputFormat.setInputPathFilter(job, BatchXMLFileFilter.class);
+    conf.set(XMLInputFormat.XML_INPUT_NAME_CONFIG, config.path);
+    context.setInput(Input.of(config.referenceName, new SourceInputFormatProvider(XMLInputFormat.class, conf)));
+  }
+
+  /**
+   * Method to set file tracking information.
+   * @param context
+   * @param conf
+   * @throws IOException
+   */
+  private void setFileTrackingInfo(BatchSourceContext context, Configuration conf) throws IOException {
+    String tableName = this.config.tableName;
+    if (StringUtils.isNotEmpty(tableName)) {
+      try {
+        processedFileTrackingTable = context.getDataset(tableName);
+        if (processedFileTrackingTable != null) {
+          List<String> processedFiles =  new ArrayList<String>();
+          File tempFile = new File(tableName);
+          // if file exists, then delete it and create new.
+          if (tempFile.exists()) {
+            tempFile.delete();
+          }
+          tempFile.createNewFile();
+          CloseableIterator<KeyValue<byte[], byte[]>> iterator = processedFileTrackingTable.scan(null, null);
+
+          //write file tracking information to temporary file.
+          while (iterator.hasNext()) {
+            KeyValue<byte[], byte[]> keyValue = iterator.next();
+            processedFiles.add(new String(keyValue.getKey(), Charsets.UTF_8));
+          }
+          conf.set(XMLInputFormat.XML_INPUTFORMAT_PROCESSED_DATA_TEMP_FILE, tableName);
+          if (config.reprocessingReq.equalsIgnoreCase("NO")) {
+            conf.set(XMLInputFormat.XML_INPUTFORMAT_PROCESSED_FILES,
+                     GSON.toJson(processedFiles, ARRAYLIST_PREPROCESSED_FILES));
+          }
+        }
+      } catch (DatasetInstantiationException datasetInstantiationException) {
```

Comment:
where is this thrown?
1. it can be propagated as is instead of converting to a runtime exception, since prepareRun allows you to throw an exception, right?
2. if it must be caught, let's catch only the statements that throw the exception, instead of the whole block